math/src/elliptic_curve/short_weierstrass/curves/bls12_381/pairing.rs

```diff
@@ -0,0 +1,270 @@
+        if *bit {
+            f = f * line(&r, q, p);
+            r = r.operate_with(q);
+            if !r.is_neutral_element() {
+                r = r.to_affine();
+            }
+        }
+    }
+    f.inv()
+}
+
+/// Auxiliary function for the final exponentiation of the ate pairing.
+fn frobenius_square(
+    f: &FieldElement<Degree12ExtensionField>,
+) -> FieldElement<Degree12ExtensionField> {
+    let [a, b] = f.value();
+    let w_raised_to_p_squared_minus_one = FieldElement::<Degree6ExtensionField>::new_base("1a0111ea397fe699ec02408663d4de85aa0d857d89759ad4897d29650fb85f9b409427eb4f49fffd8bfd00000000aaad");
+    let omega_3 = FieldElement::<Degree2ExtensionField>::new_base("1a0111ea397fe699ec02408663d4de85aa0d857d89759ad4897d29650fb85f9b409427eb4f49fffd8bfd00000000aaac");
+    let omega_3_squared = FieldElement::<Degree2ExtensionField>::new_base(
+        "5f19672fdf76ce51ba69c6076a0f77eaddb3a93be6f89688de17d813620a00022e01fffffffefffe",
+    );
+
+    let [a0, a1, a2] = a.value();
+    let [b0, b1, b2] = b.value();
+
+    let f0 = FieldElement::new([a0.clone(), a1 * &omega_3, a2 * &omega_3_squared]);
+    let f1 = FieldElement::new([b0.clone(), b1 * omega_3, b2 * omega_3_squared]);
+
+    FieldElement::new([f0, f1 * w_raised_to_p_squared_minus_one])
+}
+
+// To understand more about how to reduce the final exponentiation
+// read "Efficient Final Exponentiation via Cyclotomic Structure for
+// Pairings over Families of Elliptic Curves" (https://eprint.iacr.org/2020/875.pdf)
+//
+// TODO: implement optimizations for the hard part of the final exponentiation.
+#[allow(unused)]
+fn final_exponentiation(
+    base: &FieldElement<Degree12ExtensionField>,
+) -> FieldElement<Degree12ExtensionField> {
+    const PHI_DIVIDED_BY_R: UnsignedInteger<20> = UnsignedInteger::from("f686b3d807d01c0bd38c3195c899ed3cde88eeb996ca394506632528d6a9a2f230063cf081517f68f7764c28b6f8ae5a72bce8d63cb9f827eca0ba621315b2076995003fc77a17988f8761bdc51dc2378b9039096d1b767f17fcbde783765915c97f36c6f18212ed0b283ed237db421d160aeb6a1e79983774940996754c8c71a2629b0dea236905ce937335d5b68fa9912aae208ccf1e516c3f438e3ba79");
+
+    let f1 = base.conjugate() * base.inv();
+    let f2 = frobenius_square(&f1) * f1;
+    f2.pow(PHI_DIVIDED_BY_R)
+}
+
+/// Compute the ate pairing between point `p` in G1 and `q` in G2.
+#[allow(unused)]
+fn ate(
+    p: &ShortWeierstrassProjectivePoint<BLS12381Curve>,
+    q: &ShortWeierstrassProjectivePoint<BLS12381TwistCurve>,
+) -> FieldElement<Degree12ExtensionField> {
+    batch_ate(&[(p, q)])
+}
+
+/// Compute the product of the ate pairings for a list of point pairs.
+fn batch_ate(
+    pairs: &[(
+        &ShortWeierstrassProjectivePoint<BLS12381Curve>,
+        &ShortWeierstrassProjectivePoint<BLS12381TwistCurve>,
+    )],
+) -> FieldElement<Degree12ExtensionField> {
+    let mut result = FieldElement::one();
+    for (p, q) in pairs {
+        result = result * miller(q, p);
+    }
+    final_exponentiation(&result)
+}
```

Comment:
These functions should be public, right ?

Reply:
done! Thanks